metaflow/plugins/kfp/kfp.py

```diff
@@ -105,6 +106,61 @@ def _get_retries(node):
 
         return max_user_code_retries, max_user_code_retries + max_error_retries
 
+    @staticmethod
+    def _get_resource_requirements(node):
+        """
+        Get resource request or limit for a Metaflow step (node).
+        If @resources is also present, the maximum value is used for resource request, and lowest
+        value is used for resource limit.
+
+        Eventually resource request and limits link back to kubernetes, see
+        https://kubernetes.io/docs/concepts/configuration/manage-resources-containers/
+
+        Supported parameters: 'cpu', 'cpu_limit', 'gpu', 'gpu_vendor', 'memory', 'memory_limit'
+        Keys with no suffix set resource request (minimum);
+        keys with 'limit' suffix sets resource limit (maximum).
+
+        Default unit for memory is megabyte.
+
+        Example using resource decorator:
+            @resource(cpu=0.5, cpu_limit=2, gpu=1, memory=300)
```

Comment:
Where does the `resources` function actually get defined? Did some quick digging around the codebase and didn't find a `def resources`, is there some weird metaprogramming Metaflow does to generate that symbol?

Reply:
In each decorator class you define its name, as suggested by base StepDecorator doc string: https://github.com/zillow/metaflow/blob/master/metaflow/decorators.py#L152
It's then "imported" here: https://github.com/zillow/metaflow/blob/master/metaflow/decorators.py#L414

Reply:
Haha classic Metaflow. Definitely learned some cool metaprogramming techniques just from reading the MF codebase but these "simplifications" seem a little crazy to me!